minio/minio
# Data update tracking

## Description

By monitoring PUT/DELETE and heal operations it is possible to track changed paths and keep a bloom filter for this data. This can help prioritize paths to scan.

The bloom filter can identify paths that *have not changed*, and the few collisions will only result in a marginal extra workload. This can be implemented on either a bucket+(1 prefix level) with reasonable performance.

The bloom filter is set to have a false positive rate at 1% at 1M entries. A bloom table of this size is about ~2500 bytes when serialized.

To not force a full scan of all paths that have changed `cycle` bloom filters would need to be kept, so we guarantee that dirty paths have been scanned within `cycle` runs. Until `cycle` bloom filters have been collected all paths are considered dirty.

Bloom filters have the advantage they can be created in a distributed manner and merged later, so there is no need for a centralized bloom filter. This allows servers to independently keep bloom filters and no inter-server communication is needed.

When starting scans the bloom filters will need to be collected and reset consistently. When collecting bloom filters the merged bloom filter should be saved and kept around in case of an interrupted operation. Individual servers are responsible for persisting the bloom filter on a regular basis and on shutdown. Each server has a single global bloom filter across all sets.

## Bloom filter management

The usage scan leader will also be in charge of keeping track of bloom filter global state. 

The global state is index `X` and `Y`. 

* `Y` is the index of the current scan. This is the bloom index that is currently being written to by all servers.
* `X` is the oldest bloom index the crawler master is interested in. Typically this will be `Y-cycles`.

```
 * Scan start. 
 * Each server returns merged bloom filter after index X until (but not including) index Y.
 * Each server starts a new bloom filter with index Y if it doesn't exist (likely).
 * (Each server can dump bloom filters up to index X)
 * Scan runs using the merged bloom filters.
 * On successful scan main server increments the global index X and lower index Y.
```

This ensures that scans are resumable and that servers will be kept in sync. It also ensures that calls to switch bloom filters are idempotent, since calls with index `Y` will just continue to add data to the existing bloom filter.

As an error recovery mechanism the servers should be able to return the current `Y`, so a master server can recovers its state if it should be lost.

## Motivation and Context

Avoid crawling buckets and paths that have not changed.

## How to test this PR?

Test with `MINIO_DISK_USAGE_CRAWL_DEBUG=on MINIO_DISK_USAGE_CRAWL_DELAY=0` this will enable crawling at the fastest possible speed.

Until 16 cycles have completed the output will say `updateUsage: Start crawling. Bloom filter: false`, since there isn't complete data available yet. This should change to `updateUsage: Start crawling. Bloom filter: true` once there is enough data. It should then begin displaying `data-usage: Skipping non-updated folder: klauspost` for buckets/paths that have not been updated in 16 cycles.

Uploading files will mark paths as dirty. for example `λ mc cp mc.exe local2/dev/zPLET/2/3`

Will result in the server writing:
```
dataUpdateTracker:marking path dirty:dev
dataUpdateTracker:marking path dirty:dev/zPLET
dataUpdateTracker:marking path dirty:dev/zPLET/2
```

Which will cause these paths to be crawled for 16 cycles. Anything in `dev` that isn't `ZPLET` will still be skipped unconditionally.

The bloom filters will reset every 1000 runs which will mark all paths dirty. If data is lost all paths are considered dirty.

* Bloom filters are stored on all disks of each server in `.minio.sys/buckets/.tracker.bin`.
* The bloom cycle for the cluster is stored in `.minio.sys/buckets/.bloomcycle.bin` (erasure coded)


## Types of changes

## Checklist:
- [x] Remove debug statements with `TODO`.
- [x] Change cycle delay back from 5 seconds.
